management-node/server.go

```diff
@@ -0,0 +1,530 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/bgzzz/go-schedule/common"
+	pb "github.com/bgzzz/go-schedule/proto"
+	"github.com/bgzzz/go-schedule/wrpc"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+// ManagementNode main structure holding
+// management node grpc server logic
+type ManagementNode struct {
+	// workerNodePool is sync map of WorkerNodes
+	// that are active and connected to management
+	// node
+	workerNodePool    map[string]*wrpc.WorkerRPCClient
+	workerNodePoolMtx sync.RWMutex
+
+	// etcd is link to db
+	etcd *clientv3.Client
+
+	// scheduledTasks is sync map of tasks that are
+	// scheduled at the moment
+	scheduledTasks    map[string]*Task
+	scheduledTasksMtx sync.RWMutex
+}
+
+// Etcd prefixes that are used for storing the objects to
+// etcd
+const (
+	EtcdWorkerPrefix = "worker_"
+	EtcdTaskPrefix   = "task_"
+)
+
+// AddWorkerNode thread safely adds worker node to worker pool and etcd
+// errors in case of data inconsistency
+func (mn *ManagementNode) AddWorkerNode(wn *wrpc.WorkerRPCClient) error {
+	mn.workerNodePoolMtx.Lock()
```

Comment:
It is generally not a very good idea to grab a lock on an I/O operation (database call in this case). Any way to avoid this?

Reply:
Have change it for add/rm worker. Now, no mutexes over db IO operations.